### ISSUE TITLE
Get rid of the overhead use of pointers in field setters and `repeated` fields

### ISSUE THREAD

**Issue by cthulhu-rider** [nspcc-dev/neofs-api-go#376]
Currently we use pointers in `nested` and `repeated` message structures. With seemingly useful savings in stack memory, we thus increase the cost of garbage collection, because binding by pointer loads the heap. From the other side we use pointers where there is no memory modification, which is wrong ideologically. I propose to work with nested/repeated messages by value everywhere.

Additionally we can somewhere reduce the number of message types by flattening top-level messages when their nested structures they are not usable by themselves (e.g. `body` messages).

**Comment by fyrchik** [comment]
Good idea! I think it will definitely help in some cases and also provide a better API.

However, I would like to see some benchmarks _before_ this change. GC is only one of the factors here, others are memory locality, API consistency, copying overhead.

It is not obvious how to work with nested structs if corresponding proto-field is missing. If it can be missing, it either should be a pointer or have an `Empty` method.

So instead of doing this for all repeated values, I think we should agree on a set of "copy types", which are always used without a pointer. `ObjectID` and `Version` are good candidates. Having identifiers and attributes without pointers makes sense, but I expect `[]*Object`  (as a rather contrived example) to be better than `[]Object` by virtually any metric.

Possible benchmarks:
1. Marshaling/unmarshaling speed.
2. Stress-test specifically for marshaling/unmarshaling memory consumption (unmarshal lots of slices and drop them right after).
3. Iteration over the slice and performing read-only and/or modify actions.
4. Ad-hoc benchmarks for specific types (e.g. iterate over object attributes and cache them in map by type, then measure access time).

**Comment by cthulhu-rider** [comment]
We can test performance on several most "hot" messages (e.g. `Object`, `Container`, some request/response).

**Comment by fyrchik** [comment]
I have removed a pointer from object attributes slice and results are very good so far (50 attributes, each key ~20 bytes).

```
name                           old time/op    new time/op    delta
AttributesMarshal/marshal-8      4.74µs ±19%    2.58µs ± 2%  -45.65%  (p=0.000 n=10+10)
AttributesMarshal/unmarshal-8    3.13µs ± 7%    0.55µs ± 1%  -82.30%  (p=0.000 n=9+8)

name                           old alloc/op   new alloc/op   delta
AttributesMarshal/marshal-8      4.42kB ± 0%    4.42kB ± 0%     ~     (all equal)
AttributesMarshal/unmarshal-8    2.02kB ± 0%    1.79kB ± 0%  -11.11%  (p=0.000 n=10+10)

name                           old allocs/op  new allocs/op  delta
AttributesMarshal/marshal-8        51.0 ± 0%      51.0 ± 0%     ~     (all equal)
AttributesMarshal/unmarshal-8      51.0 ± 0%       1.0 ± 0%  -98.04%  (p=0.000 n=10+10)
```

**Comment by fyrchik** [comment]
Iteration over the attributes (simple for-loop, function either accepts pointer or not):
```
name                                       old time/op  new time/op  delta
AttributesMarshal/iteration,_pointer-8     26.7ns ± 1%  26.5ns ± 0%  -0.51%  (p=0.008 n=9+9)
AttributesMarshal/iteration,_no_pointer-8  26.8ns ± 3%  26.7ns ± 1%    ~     (p=0.867 n=10+10)
```
